bin/GoogleAdsCleanupServices.php

```diff
@@ -0,0 +1,529 @@
+		$this->remove_file( "/{$this->version}/Services/{$service}Service.php" );
+		$this->remove_file( "/{$this->version}/Resources/{$service}.php" );
+	}
+
+	/**
+	 * Find used enums and remove any unused ones.
+	 */
+	protected function remove_enums() {
+		$library_enums = $this->find_library_enums();
+		$used_enums    = $this->find_used_enums();
+
+		foreach ( array_diff( $library_enums, $used_enums ) as $enum ) {
+			$this->remove_enum( $enum );
+		};
+	}
+
+	/**
+	 * Remove a specific enum.
+	 *
+	 * @param string $enum Enum name.
+	 */
+	protected function remove_enum( string $enum ) {
+		$this->output_text( "Removing enum {$enum}" );
+		$this->src_path = '/src/Google/Ads/GoogleAds';
+
+		$this->remove_file( "/{$this->version}/Enums/{$enum}Enum.php" );
+		$this->remove_file( "/{$this->version}/Enums/{$enum}Enum_{$enum}.php" );
+		$this->remove_file( "/{$this->version}/Enums/{$enum}Enum/{$enum}.php" );
+		$this->remove_directory( "/{$this->version}/Enums/{$enum}Enum" );
+
+		// Remove metadata files.
+		$this->src_path = '/metadata/Google/Ads/GoogleAds';
+		$this->remove_file( "/{$this->version}/Enums/{$enum}.php" );
+	}
+
+	/**
+	 * Find enum names within the library.
+	 *
+	 * @return array
+	 */
+	protected function find_library_enums(): array {
+		$path = "{$this->path}/metadata/Google/Ads/GoogleAds/{$this->version}/Enums";
+
+		$output = glob( "{$path}/*.php" );
+
+		if ( empty( $output ) ) {
+			return [];
+		}
+
+		return array_map(
+			function( $file ) {
+				preg_match( '/(.*)\.php/', basename( $file ), $matches );
+				return $matches[1];
+			},
+			$output
+		);
+	}
+
+	/**
+	 * Find enum names used within the extension.
+	 *
+	 * @return array
+	 */
+	protected function find_used_enums(): array {
+		$pattern = "use Google\\\\Ads\\\\GoogleAds\\\\{$this->version}\\\\Enums\\\\([A-Za-z]+)Enum\\\\";
+		$command = "grep -rE --include=*.php '{$pattern}' {$this->code_path}";
+
+		// phpcs:ignore WordPress.PHP.DiscouragedPHPFunctions.system_calls_exec
+		exec( $command, $output );
+
+		if ( empty( $output ) ) {
+			return [];
+		}
+
+		return array_map(
+			function( $line ) use ( $pattern ) {
+				preg_match( "/{$pattern}/", $line, $matches );
+				return $matches[1];
+			},
+			$output
+		);
+	}
+
+	/**
+	 * Remove a use statement from a file.
+	 *
+	 * @param string $file
+	 * @param string $name
+	 */
+	protected function remove_use_statement( string $file, string $name ) {
+		$pattern = '/\s*use\s+[\w\d\\\]+' . preg_quote( $name, '/' ) . ';/';
+		$this->remove_pattern( $file, $pattern );
+	}
+
+	/**
+	 * Remove a function from a file.
+	 *
+	 * @param string $file
+	 * @param string $name Function name.
+	 */
+	protected function remove_function( string $file, string $name ) {
+		$file = $this->file_path( $file );
+		if ( ! file_exists( $file ) ) {
+			$this->warning( sprintf( 'File does not exist: %s', $file ) );
+			return;
+		}
+
+		$contents = file_get_contents( $file ); // phpcs:ignore WordPress.WP.AlternativeFunctions
```

Comment:
❓ 

It's necessary to remove those 2 functions?

Reply:
So my worries are about how safe is this function  and how potentially can create problems parsing and deleting text or code we don't control that could change over time. 

For example introducing a bracket in a comment like inside the function body (just a random example).

Since post install script is not failing for using an undefined class inside a never invoked function and since this script is for reducing space rather than cleaning the Google Code I was wondering if its worth it to remove those functions.

But was just an observation and it's not a blocker.

Reply:
I agree that those functions don't cover us 100% and there are flaws. However if I look at the bigger picture, I do think it's still worth it to continue removing them.

Initially we asked the owners of the library for a script to remove unused services, there was desire to implement it but it didn't have a high priority. So this script is our interim solution to slim down the library solely for our use case. It never set out to be a complete solution.

Even though we don't control the code, updating the library is a manual process which is done through a PR because we override classes to make use of the library. So if any part of this script would break a function for future versions of the library, it would be caught when running the unit tests in the PR and it won't go unnoticed.